beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs

```diff
@@ -53,7 +54,8 @@ use types::{BlobSidecar, Epoch, EthSpec, Hash256};
 pub struct PendingComponents<T: EthSpec> {
     pub block_root: Hash256,
     pub verified_blobs: FixedVector<Option<KzgVerifiedBlob<T>>, T::MaxBlobsPerBlock>,
-    pub executed_block: Option<AvailabilityPendingExecutedBlock<T>>,
+    /// FIXME: change this back to `AvailabilityPendingExecutedBlock` once tree-states is merged
```

Comment:
Hmmm this suggests an alternative design. Perhaps we get rid of `DietAvailabilityPendingExecutedBlock` and change  [`BlockImportData`](https://github.com/sigp/lighthouse/blob/c7ddf1f0b1cfc19a03fcadc7d5ab13d6d595c69d/beacon_node/beacon_chain/src/block_verification_types.rs#L291) to include an `Option<BeaconState>`. This creates the hybrid type you mentioned. Then when serializing we would always serialize as `None`. We would also probably move the `recover_pending_executed_block()` functionality to a method on `AvailabilityPendingExecutedBlock` instead of the `StateLRUCache`.

Then this PR can behave pretty much the same way, but we can continue to use this as you have described once `tree-states` is merged.

What do you think?

Reply:
Yeah I think that sounds good! I was trying to suggest a minimal change atop what you've already implemented, but if you're happy to do a more major re-structure then I think an `Option<BeaconState<E>>` is more future proof. Thanks!

Reply:
Okay so I realized this is a bad design since the `ImportData` should only ever be stripped of the state when inside the `DataAvailabiltyChecker` and designing it so that it could be passed around outside the cache while not containing the state forces us to write code to handle that case (which shouldn't happen) in a ton of places. Better to instead contain this complexity within the cache. I've deleted the `ssz_tagged_beacon_state` function and removed `Encode`, `Decode`, & `Clone` from `ImportData` & `AvailabilityPendingExecutedBlock` since we really shouldn't be doing those things with them (and we no longer are with this change).

Perhaps we should decide in the future how best to handle recovering from disk after tree states.. one really easy option would be to just keep the `StateLRUCache` but increase its size considerably after tree states. Keeping the diet type around and separating the state from the cache entries certainly makes [this call](https://github.com/sigp/lighthouse/blob/840907fcd3e9c1cbe789d308461cfc1454721c56/beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs#L548) much faster..

Reply:
Yeah nice, sounds good. I like the deletions in the latest diff :grin: